File: Python/프로그래머스/4/12983. 단어 퍼즐/단어 퍼즐.py
INF = float('inf')

def solution(strs, t):
    ans = 0
    dp = [INF] * (len(t)+1)
    set_strs = set(strs)
    dp[0] = 0
    
    for i in range(1, len(t)+1):
        for k in range(5):
            s_i = i-1 - k
            if (s_i < 0):                
                break
            
            if (t[s_i:i] in strs):
                if (s_i == 0):
                    dp[i] = min(dp[i], 1)
                else:
                    dp[i] = min(dp[i], dp[s_i]+1)
            
    if (dp[-1] == INF):
        return -1

    return dp[-1]

File: C++/프로그래머스/4/12983. 단어 퍼즐/단어 퍼즐.cpp
#include <bits/stdc++.h>
using namespace std;
int INF = 1000000000;
    
int solution(vector<string> strs, string t)
{
    int ans = 0;
    int len_t = t.size();
    unordered_set<string> strs_set;
    vector<int> dp(len_t+1, INF);
    dp[0] = 0;
    
    for (auto s: strs) strs_set.insert(s);
    
    
    for (int i=1; i < len_t+1; i++){
        
        for (int j=0; j < 5; j++){
            int s_idx = i-1 - j, len_w = j+1;
            if (s_idx < 0) continue;
            
            string sub = t.substr(s_idx, len_w);
            // cout << "sub: " << sub << endl;
            if (strs_set.count(sub)){
                if (s_idx == 0)
                    dp[i] = 1;
                else{
                    dp[i] = min(dp[i], dp[s_idx]+1);
                }
            }
        }
    }
    
    // for (auto n: dp)
    //     cout << n << " ";
    
    ans = dp.back();
    
    if (ans == INF) return -1;
    return ans;
}